funsor/distributions.py

```diff
@@ -451,22 +452,77 @@ def eager_mvn(loc, scale_tril, value):
 
 
 # Create a Gaussian from a ground observation.
-@eager.register(MultivariateNormal, Tensor, Tensor, Variable)
-@eager.register(MultivariateNormal, Variable, Tensor, Tensor)
+# TODO refactor this logic into Gaussian.eager_subs() and
+#   here return Gaussian(...scale_tril...)(value=loc-value).
+@eager.register(MultivariateNormal, (Variable, Contraction), Tensor, (Variable, Contraction))
+@eager.register(MultivariateNormal, (Variable, Contraction), Tensor, Tensor)
+@eager.register(MultivariateNormal, Tensor, Tensor, (Variable, Contraction))
 def eager_mvn(loc, scale_tril, value):
-    if isinstance(loc, Variable):
-        loc, value = value, loc
+    affine = loc - value
+    if not isinstance(affine, Contraction):
+        return None  # lazy
+
+    const, coeffs = extract_affine(affine)
+    if not isinstance(const, Tensor):
+        return None  # lazy
+    if not all(isinstance(coeff, Tensor) for coeff, _ in coeffs.values()):
+        return None  # lazy
+
+    # Dovetail to avoid variable name collision in einsum.
+    equations1 = [''.join(c if c in '.,->' else chr(ord(c) * 2))
+                  for _, eqn in coeffs.values() for c in eqn]
+    equations2 = [''.join(c if c in '.,->' else chr(ord(c) * 2 + 1))
+                  for _, eqn in coeffs.values() for c in eqn]
 
-    dim, = loc.output.shape
-    inputs, (loc, scale_tril) = align_tensors(loc, scale_tril)
-    inputs.update(value.inputs)
-    int_inputs = OrderedDict((k, v) for k, v in inputs.items() if v.dtype != 'real')
+    real_inputs = OrderedDict((k, v) for k, v in affine.inputs.items() if v.dtype == 'real')
+    assert len(real_inputs) == len(coeffs)
+    tensors = [coeffs[k] for k in real_inputs] + [const, scale_tril]
+    inputs, tensors = align_tensors(*tensors)
 
-    precision = cholesky_inverse(scale_tril)
+    # Incorporate scale_tril before broadcasting.
+    scale_tril = tensors[-1]
+    prec_sqrt = cholesky_inverse(scale_tril)  # FIXME is transpose correct?
+    for i in range(len(coeffs)):
+        coeff = tensors[i]
+        inputs1, output1 = equations1[i].split('->')
+        input11, _ = inputs1.split(',')
+        inputs2, output2 = equations2[i].split('->')
+        input21, _ = inputs2.split(',')
+        coeff = torch.einsum(f'...{input11},...{output1}{output2}->...{input21}',
```

Comment:
super interesting! I am not sure if this is {output1}{output2} or {output2}{output1} (probably depends on the definition of prec_sqrt - i.e. if transpose is involved). In addition, if we do not broadcast `scale_tril`, I guess `torch.einsum` will fail here because Ellipses of coeff and prec_sqrt are mismatched.

Reply:
Good point, maybe I can leverage the new `Einsum` logic from #249 ...